compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/cfa/ControlFlowInfos.kt

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.cfa
+
+import kotlinx.collections.immutable.PersistentMap
+import kotlinx.collections.immutable.persistentMapOf
+import org.jetbrains.kotlin.contracts.description.EventOccurrencesRange
+import org.jetbrains.kotlin.fir.resolve.dfa.cfg.CFGNode
+import org.jetbrains.kotlin.fir.resolve.dfa.cfg.EdgeLabel
+import org.jetbrains.kotlin.fir.resolve.dfa.cfg.NormalPath
+import org.jetbrains.kotlin.fir.resolve.dfa.cfg.TryExpressionExitNode
+import org.jetbrains.kotlin.fir.symbols.FirBasedSymbol
+import org.jetbrains.kotlin.fir.symbols.impl.FirPropertySymbol
+import java.lang.IllegalStateException
+
+// TODO: this can be `sealed` once all of its subtypes are migrated
+abstract class ControlFlowInfo<S : ControlFlowInfo<S, K, V>, K : Any, V : Any> protected constructor(
+    protected val map: PersistentMap<K, V>,
+) : PersistentMap<K, V> by map {
+
+    protected abstract val constructor: (PersistentMap<K, V>) -> S
+
+    protected abstract val empty: () -> S
+
+    override fun equals(other: Any?): Boolean {
+        return map == (other as? ControlFlowInfo<*, *, *>)?.map
+    }
+
+    override fun hashCode(): Int {
+        return map.hashCode()
+    }
+
+    override fun toString(): String {
+        return map.toString()
+    }
+
+    override fun put(key: K, value: V): S {
+        return constructor(map.put(key, value))
+    }
+
+    abstract fun merge(other: S): S
+}
+
+private fun <S : ControlFlowInfo<S, K, EventOccurrencesRange>, K : Any> merge(one: S, other: S): S {
+    var result = one
+    for (symbol in one.keys.union(other.keys)) {
+        val kind1 = one[symbol] ?: EventOccurrencesRange.ZERO
+        val kind2 = other[symbol] ?: EventOccurrencesRange.ZERO
+        result = result.put(symbol, kind1 or kind2)
+    }
+    return result
+}
+
+class LambdaInvocationInfo(
```

Comment:
According to previous comment, please move this class back to `FirCallsEffectAnalyzer.kt`

Reply:
Rewrote commits to not move this and other classes in the beginning.